artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonAbstractReceiver.java

```diff
@@ -345,25 +345,23 @@ public void onMessageComplete(Delivery delivery,
                           Message message, DeliveryAnnotations deliveryAnnotations) {
       connection.requireInHandler();
 
-      try {
-         receiver.advance();
-
-         Transaction tx = null;
-         if (delivery.getRemoteState() instanceof TransactionalState) {
-            TransactionalState txState = (TransactionalState) delivery.getRemoteState();
-            try {
-               tx = this.sessionSPI.getTransaction(txState.getTxnId(), false);
-            } catch (Exception e) {
-               this.onExceptionWhileReading(e);
-            }
-         }
+      receiver.advance();
 
-         actualDelivery(message, delivery, deliveryAnnotations, receiver, tx);
-      } finally {
-         // reader is complete, we give it up now
-         this.messageReader.close();
-         this.messageReader = null;
+      Transaction tx = null;
+      if (delivery.getRemoteState() instanceof TransactionalState) {
+         TransactionalState txState = (TransactionalState) delivery.getRemoteState();
+         try {
+            tx = this.sessionSPI.getTransaction(txState.getTxnId(), false);
+         } catch (Exception e) {
+            this.onExceptionWhileReading(e);
+         }
       }
+
+      //  messageReader.close should be called before the delivery, to signal the reader is complete.
+      this.messageReader.close();
+      this.messageReader = null;
```

Comment:
This null assignment should remain in a finally block as it ensures that the reader is always nulled when onMessageComplete gets called.